### PR TITLE
infra controller for cluster-api

### PR DESCRIPTION
**Pull request by staebler**
This modifies the infra controller so that it can be used with the cluster-api resources as well.

The result of syncing infra controller on a `Cluster` from cluster-api is a `ProviderStatus` filled out in the `Cluster` like so.

```
    providerStatus:
      apiVersion: clusteroperator.openshift.io/v1alpha1
      conditions:
      - lastProbeTime: 2018-05-10T18:26:32Z
        lastTransitionTime: 2018-05-10T18:26:32Z
        message: Job completed successful
        reason: JobCompleted
        status: "True"
        type: InfraProvisioned
      kind: ClusterProviderStatus
      machineSetCount: 0
      metadata:
        creationTimestamp: null
      provisioned: true
      provisionedJobGeneration: 1
      ready: false
```

### CODE REVIEW

File: pkg/apis/clusteroperator/v1alpha1/types.go [hunk 1]
```diff
@@ -658,3 +660,18 @@ const (
 	// NodeTypeCompute is a node that is a compute node in the cluster
 	NodeTypeCompute NodeType = "Compute"
 )
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CombinedCluster is a cluster object that combines the information from
+// cluster-operator Cluster and cluster-api Cluster.
+type CombinedCluster struct {
```
Comment (dgoodwin) [inline pkg/apis/clusteroperator/v1alpha1/types.go]: Is this type temporary?

Comment (staebler) [reply pkg/apis/clusteroperator/v1alpha1/types.go]: Not really. We need some way of keeping the decoded ProviderConfig and ProviderStatus around while we work on the Cluster. So I would expect that we will finalize on something similar to this even after we remove Cluster from clusteroperator.openshift.io.

Comment (dgoodwin) [reply pkg/apis/clusteroperator/v1alpha1/types.go]: Ok struck me as a little odd given the ClusterAPI* structs already contain the other ClusterOperator* portions, just not decoded.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -309,10 +312,10 @@ func MachineSetLabels(cluster *clusteroperator.Cluster, machineSetShortName stri
 
 // JobLabelsForClusterController returns the labels to apply to a job doing a task
 // for the specified cluster.
-func JobLabelsForClusterController(cluster *clusteroperator.Cluster, jobType string) map[string]string {
+func JobLabelsForClusterController(cluster metav1.Object, jobType string) map[string]string {
```
[the remark this answers is not in the document]
Comment (staebler) [reply pkg/controller/controller_utils.go]: Done.

**Comment by staebler** [comment]
It should be noted that the workaround for `ClusterVersion` will only work if the `ClusterVersion` exists prior to the infra controller attempting to sync the `Cluster`.

File: pkg/controller/controller_utils_test.go [hunk 3]
```diff
@@ -1166,8 +1166,34 @@ kind: Machine
...
 		})
 	}
 }
+
+func TestClusterAPIProviderStatusFromClusterStatus(t *testing.T) {
+	clusterStatus := &clusteroperator.ClusterStatus{
+		MachineSetCount:      1,
+		MasterMachineSetName: "master",
+	}
+	//expectedProviderStatus: `{"apiVersion":"clusteroperator.openshift.io/v1alpha1","kind":"ClusterProviderStatus","machineSetCount":1,"masterMachineSetName":"master"}`,
```
Comment (dgoodwin) [inline pkg/controller/controller_utils_test.go]: Can remove now.

**Comment by dgoodwin** [comment]
This looks fine to me, I am going to merge as I would like to build on the status updating code. If @csrwng has any feedback we can do a followup PR.

/lgtm

**Pull request merged**